### PR TITLE
ovirt: imageio volume populator for cold migration

### PR DESCRIPTION
**Pull request by bennyz**
Used with https://github.com/kubev2v/ovirt-imageio-populator

- [x] Handle access modes/volume mode - we may want to reuse StorageProfile for this
- [x] Handle fsOverhead
- [x] Handle deployment - done in seperate PR
- [x] Error handling - hopefully completed. cleanup required on separate PR.
- [x] Integrate status update (ovirt-imageio-populator side ready in https://github.com/kubev2v/ovirt-imageio-populator/pull/8)
- [x] ~~Figure out VolumeAccess~~
- [x] ~~Permissions (ClusterRole might be needed)~~

### CODE REVIEW

File: pkg/controller/plan/migration.go [hunk 1]
```diff
@@ -916,7 +954,118 @@ func (r *Migration) resetPrecopyTasks(vm *plan.VMStatus, step *plan.Step) {
 	}
 }
 
-//
+func (r *Migration) createVolumes(vm ref.Ref) (err error) {
```
Comment (bennyz) [inline pkg/controller/plan/migration.go:957]: move to ovirt/builder.go

File: pkg/controller/plan/migration.go [hunk 2]
```diff
@@ -916,7 +954,118 @@ func (r *Migration) resetPrecopyTasks(vm *plan.VMStatus, step *plan.Step) {
...
+					"namespace": r.Plan.Spec.TargetNamespace,
+				},
+				"spec": map[string]interface{}{
+					"storageClassName": storageName,
+					"resources": map[string]interface{}{
+						"requests": map[string]interface{}{
+							"storage": resource.NewQuantity(da.Disk.ProvisionedSize, resource.BinarySI).String(),
+						},
+					},
+					"accessModes": []string{"ReadWriteOnce"},
```
Comment (bennyz) [inline pkg/controller/plan/migration.go:1009]: Not necessarily correct

File: pkg/controller/plan/migration.go [hunk 3]
```diff
@@ -634,6 +634,18 @@ func (r *Migration) execute(vm *plan.VMStatus) (err error) {
 		vm.Phase = r.next(vm.Phase)
 	case ScheduleVM:
 		step, found := vm.FindStep(r.step(vm))
+
+		// Skip VM start
+		if *r.Plan.Provider.Source.Spec.Type == v1beta1.OVirt {
```
Comment (bennyz) [inline pkg/controller/plan/migration.go:639]: Not needed, will be removed entirely

File: pkg/controller/plan/migration.go [hunk 4]
```diff
@@ -916,7 +954,118 @@ func (r *Migration) resetPrecopyTasks(vm *plan.VMStatus, step *plan.Step) {
...
+	if err != nil {
+		return
+	}
+	url, err := url.Parse(r.Source.Provider.Spec.URL)
+	if err != nil {
+		return
+	}
+	storageName := &r.Context.Map.Storage.Spec.Map[0].Destination.StorageClass
+	for _, da := range ovirtVm.DiskAttachments {
+		populatorCr := v1beta1.OvirtImageIOPopulator{
```
Comment (bennyz) [inline pkg/controller/plan/migration.go:975]: add ownerReference

File: pkg/controller/plan/migration.go [hunk 5]
```diff
@@ -916,7 +954,118 @@ func (r *Migration) resetPrecopyTasks(vm *plan.VMStatus, step *plan.Step) {
...
+		r.Log.Info("Creating populator CR", "CR", populatorCr)
+		err = r.Client.Create(context.Background(), &populatorCr, &client.CreateOptions{})
+		if err != nil {
+			return
+		}
+		apiGroup := "forklift.konveyor.io"
+
+		pvc := &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"kind":       "PersistentVolumeClaim",
```
Comment (bennyz) [inline pkg/controller/plan/migration.go:996]: add OwnerReference

File: pkg/apis/forklift/v1beta1/BUILD.bazel [hunk 6]
```diff
@@ -8,6 +8,7 @@ go_library(
         "host.go",
         "mapping.go",
         "migration.go",
+        "ovirtimagevolumepopulator.go",
```
Comment (ahadas) [inline pkg/apis/forklift/v1beta1/BUILD.bazel:11]: maybe it better be `pkg/apis/forklift/v1beta1/populator/ovirt-imageio.go`?

File: pkg/controller/plan/migration.go [hunk 7]
```diff
@@ -580,6 +573,13 @@ func (r *Migration) execute(vm *plan.VMStatus) (err error) {
 		}
 	case CreateDataVolumes:
 		step, found := vm.FindStep(r.step(vm))
+		if *r.Plan.Provider.Source.Spec.Type == v1beta1.OVirt {
+			if vm.Warm == nil {
```
Comment (ahadas) [inline pkg/controller/plan/migration.go:577]: yeah, not warm and also if the target is the 'host' provider

File: pkg/controller/plan/migration.go [hunk 8]
```diff
@@ -580,6 +573,13 @@ func (r *Migration) execute(vm *plan.VMStatus) (err error) {
 		}
 	case CreateDataVolumes:
 		step, found := vm.FindStep(r.step(vm))
+		if *r.Plan.Provider.Source.Spec.Type == v1beta1.OVirt {
```
Comment (ahadas) [inline pkg/controller/plan/migration.go:576]: if we don't look at 'step' then we can skip 575

File: pkg/controller/plan/migration.go [hunk 9]
```diff
@@ -916,7 +943,124 @@ func (r *Migration) resetPrecopyTasks(vm *plan.VMStatus, step *plan.Step) {
...
+				EngineSecretName: r.Source.Secret.Name,
+				DiskID:           da.Disk.ID,
+			},
+		}
+
+		err = r.Client.Create(context.Background(), &populatorCr, &client.CreateOptions{})
+		if err != nil {
+			return
+		}
+		apiGroup := "forklift.konveyor.io"
```
Comment (ahadas) [inline pkg/controller/plan/migration.go:981]: we can take it from [here](https://github.com/kubev2v/forklift/blob/main/pkg/apis/forklift/v1beta1/register.go#L33)

File: pkg/controller/plan/migration.go [hunk 10]
```diff
@@ -1170,7 +1310,50 @@ func (r *Migration) waitingForFirstConsumer(vm *plan.VMStatus) (waiting bool, er
...
+		r.vmMap, err = r.kubevirt.VirtualMachineMap()
+		if err != nil {
+			return
+		}
+	}
+
+	var vmCr VirtualMachine
+	found := false
+	if vmCr, found = r.vmMap[vm.ID]; !found {
+		msg := "VirtualMachine CR not found."
```
Comment (ahadas) [inline pkg/controller/plan/migration.go:1324]: how can this happen, when someone modifies the migration CR?

**Review by ahadas — COMMENTED** [review]
nice!

File: pkg/controller/plan/migration.go (hunk 10 above)
Comment (bennyz) [reply pkg/controller/plan/migration.go:1324]: I'm not sure to be honest, it's a copy paste, it may not be actually needed here

File: pkg/controller/plan/migration.go [hunk 11]
```diff
@@ -916,7 +943,124 @@ func (r *Migration) resetPrecopyTasks(vm *plan.VMStatus, step *plan.Step) {
...
+
+	for _, da := range ovirtVm.DiskAttachments {
+		obj := client.ObjectKey{Namespace: r.Plan.Spec.TargetNamespace, Name: da.Disk.ID}
+		pvc := core.PersistentVolumeClaim{}
+		err = r.Client.Get(context.Background(), obj, &pvc)
+		if err != nil {
+			return
+		}
+
+		if pvc.Status.Phase != core.ClaimBound {
```
Comment (bennyz) [inline pkg/controller/plan/migration.go:1046]: may make sense to stop and error on `core.ClaimLost`

File: pkg/controller/plan/migration.go [hunk 12]
```diff
@@ -916,7 +943,124 @@ func (r *Migration) resetPrecopyTasks(vm *plan.VMStatus, step *plan.Step) {
...
+			},
+		}
+
+		err = r.Client.Create(context.Background(), &populatorCr, &client.CreateOptions{})
+		if err != nil {
+			return
+		}
+		apiGroup := "forklift.konveyor.io"
+
+		pvc := &unstructured.Unstructured{
```
Comment (liranr23) [inline pkg/controller/plan/migration.go:983]: why it's unstructured and not core.PersistentVolumeClaim?

File: pkg/controller/plan/migration.go [hunk 13]
```diff
@@ -916,7 +943,124 @@ func (r *Migration) resetPrecopyTasks(vm *plan.VMStatus, step *plan.Step) {
...
+		}
+
+		pvcResource := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "persistentvolumeclaims"}
+
+		_, err = dynamicClient.Resource(pvcResource).Namespace(r.Plan.Spec.TargetNamespace).Create(context.TODO(), pvc, meta.CreateOptions{})
+
+		if err != nil {
+			return
+		}
+	}
```
Comment (liranr23) [inline pkg/controller/plan/migration.go:1025]: if we use PVC above, do we need to use dynamic client or the client we already use will be enough?

File: pkg/controller/plan/migration.go (hunk 10 above)
Comment (liranr23) [reply pkg/controller/plan/migration.go:1324]: currently, the code is based on the idea the VM posted, then it takes details from it. With the change in https://github.com/kubev2v/forklift/pull/48 we probably won't need it.
Here below we just search for the volumes, we can query them using the client.

File: pkg/controller/plan/migration.go (hunk 12 above)
Comment (bennyz) [reply pkg/controller/plan/migration.go:983]: Because we need the dataSourceRef field which doesn't exist in the struct on the version we use, if we upgrade the client then we should switch to the typed version, I thought I had a TODO for this but it was probably in the other version

Comment (ahadas) [reply pkg/controller/plan/migration.go:983]: @bennyz as it's a repeating question, it is worth to document it..

Comment (bennyz) [reply pkg/controller/plan/migration.go:983]: added

Comment (liranr23) [reply pkg/controller/plan/migration.go]: once we will bump the code dep, we can drop this (in https://github.com/kubev2v/forklift/pull/64)

File: pkg/controller/plan/migration.go [hunk 14]
```diff
@@ -916,7 +954,118 @@ func (r *Migration) resetPrecopyTasks(vm *plan.VMStatus, step *plan.Step) {
...
+				"apiVersion": "v1",
+				"metadata": map[string]interface{}{
+					"name":      da.DiskAttachment.ID,
+					"namespace": r.Plan.Spec.TargetNamespace,
+				},
+				"spec": map[string]interface{}{
+					"storageClassName": storageName,
+					"resources": map[string]interface{}{
+						"requests": map[string]interface{}{
+							"storage": resource.NewQuantity(da.Disk.ProvisionedSize, resource.BinarySI).String(),
```
[the remark this answers is not in the document]
Comment (liranr23) [reply pkg/controller/plan/migration.go]: i added 10% hardcoded for now..

File: pkg/controller/plan/migration.go (hunk 12 above)
Comment (ahadas) [reply pkg/controller/plan/migration.go]: > once we will bump the code dep, we can drop this (in #64)

that's fine but it means we won't support OCP 4.10, right?

File: pkg/controller/plan/adapter/vsphere/builder.go [hunk 15]
```diff
@@ -690,3 +692,16 @@ func (r *Builder) host(hostID string) (host *model.Host, err error) {
 func trimBackingFileName(fileName string) string {
 	return backingFilePattern.ReplaceAllString(fileName, ".vmdk")
 }
+
+// Build an OvirtVolumePopulator for XDiskAttachment and source URL, not supported for vSphere
+func (r *Builder) OvirtVolumePopulator(da ovirt.XDiskAttachment, sourceUrl *url.URL) *api.OvirtVolumePopulator {
```
Comment (bennyz) [inline pkg/controller/plan/adapter/vsphere/builder.go:697]: We need to rethink this at some point, it's way too specific, and it forces ovirt imports in a vsphere package

Comment (liranr23) [reply pkg/controller/plan/adapter/vsphere/builder.go:697]: do you mean renaming it?

Comment (bennyz) [reply pkg/controller/plan/adapter/vsphere/builder.go:697]: not the name as we need the type here too, but perhaps it would make more sense to move it out to be a standalone function and not part of the builder interface, it doesn't depend on too many fields from `builder`

File: pkg/controller/plan/kubevirt.go [hunk 16]
```diff
@@ -555,11 +572,121 @@ func (r *KubeVirt) getPVCs(vm *plan.VMStatus) (pvcs []core.PersistentVolumeClaim
...
+				volumeMode = claimProperty.VolumeMode
+			}
+			break
+		}
+	}
+	return
+}
+
+// Return if the import done with OvirtImageIO
+func (r *KubeVirt) isOvirtImageIO(vm *plan.VMStatus) bool {
```
Comment (bennyz) [inline pkg/controller/plan/kubevirt.go:683]: Maybe `isOvirtProvider`?

File: pkg/controller/plan/migration.go (hunk 12 above)
Comment (liranr23) [reply pkg/controller/plan/migration.go]: basing on https://kubernetes.io/blog/2021/08/30/volume-populators-redesigned/ it's since k8s 1.22, OCP 4.9 uses k8s 1.22. OCP 4.10 uses 1.23.

**Comment by bennyz** [comment]
lgtm (can't approve my own PR apparently)

File: pkg/controller/plan/migration.go (hunk 12 above)
Comment (ahadas) [reply pkg/controller/plan/migration.go]: > basing on https://kubernetes.io/blog/2021/08/30/volume-populators-redesigned/ it's since k8s 1.22, OCP 4.9 uses k8s 1.22. OCP 4.10 uses 1.23.

`In v1.22 if the AnyVolumeDataSource feature gate is enabled, ..` and ask @bennyz about his experience in enabling it on such cluster..
anyway, now that OCP 4.12 is out, we don't care about 4.10 and it's enabled by default on OCP 4.11 so that's alright

File: pkg/controller/plan/kubevirt.go [hunk 17]
```diff
@@ -312,6 +315,20 @@ func (r *KubeVirt) EnsureVM(vm *plan.VMStatus) (err error) {
 			err = liberr.Wrap(err)
 			return
 		}
+		if r.isOvirtProvider(vm) && pvc.Spec.DataSource.Kind == "OvirtVolumePopulator" {
```
Comment (ahadas) [inline pkg/controller/plan/kubevirt.go:318]: I'm not sure if there's a better way to do it but how about replacing those strings with something like reflect.TypeOf(OvirtVolumePopulator{}).Name() ?

File: pkg/controller/plan/kubevirt.go [hunk 18]
```diff
@@ -555,11 +572,136 @@ func (r *KubeVirt) getPVCs(vm *plan.VMStatus) (pvcs []core.PersistentVolumeClaim
...
+		},
+		Spec: v1beta1.OvirtVolumePopulatorSpec{
+			EngineURL:        fmt.Sprintf("https://%s", sourceUrl.Host),
+			EngineSecretName: r.Source.Secret.Name,
+			DiskID:           da.Disk.ID,
+		},
+	}
+}
+
+func (r *KubeVirt) getOvirtPVCs(vm ref.Ref, step *plan.Step) (ready bool, err error) {
```
Comment (ahadas) [inline pkg/controller/plan/kubevirt.go:643]: why the name is getOvirtPVCs? it doesn't seem to return PVCs here

File: pkg/controller/plan/kubevirt.go [hunk 19]
```diff
@@ -555,11 +572,136 @@ func (r *KubeVirt) getPVCs(vm *plan.VMStatus) (pvcs []core.PersistentVolumeClaim
...
+				accessModes = append(accessModes, claimProperty.AccessModes...)
+				volumeMode = claimProperty.VolumeMode
+			}
+			break
+		}
+	}
+	return
+}
+
+// Return if the import done with OvirtVolumePopulator
```
Comment (ahadas) [inline pkg/controller/plan/kubevirt.go:697]:
```suggestion
// Return true when the import is done with OvirtVolumePopulator
```

File: pkg/controller/plan/migration.go [hunk 20]
```diff
@@ -566,6 +569,13 @@ func (r *Migration) execute(vm *plan.VMStatus) (err error) {
 			vm.AddError(fmt.Sprintf("Step '%s' not found", r.step(vm)))
 			break
 		}
+		if r.kubevirt.isOvirtProvider(vm) {
+			err = r.kubevirt.createVolumes(vm.Ref)
```
Comment (ahadas) [inline pkg/controller/plan/migration.go:573]: missing a check of the error?

File: pkg/controller/plan/migration.go [hunk 21]
```diff
@@ -1126,6 +1162,39 @@ func (r *Migration) setRunning(vm *plan.VMStatus, running bool) (err error) {
...
 
+func (r *Migration) updateCopyProgressForOvirt(vm *plan.VMStatus, step *plan.Step) (err error) {
+	pvcs, err := r.kubevirt.getPVCs(vm)
+	if err != nil {
+		return
+	}
+	for _, pvc := range pvcs {
+		claim := pvc.Spec.DataSource.Name
+		var task *plan.Task
+		found := false
```
Comment (ahadas) [inline pkg/controller/plan/migration.go:1173]: the two lines above seems redundant here as well

**Review by ahadas — COMMENTED** [review]
too many is-ovirt things now but that should hopefully be improved with the addition of more populators, tried to keep the comments on things that are easy to change quickly so we can get this in

File: pkg/controller/plan/kubevirt.go (hunk 19 above)
Comment (ahadas) [reply pkg/controller/plan/kubevirt.go:697]: or "Check if .."

File: pkg/controller/plan/migration.go (hunk 11 above)
Comment (ahadas) [reply pkg/controller/plan/migration.go]: yeah but let's consider this as part of the addition of OSP

File: pkg/controller/plan/kubevirt.go (hunk 18 above)
Comment (liranr23) [reply pkg/controller/plan/kubevirt.go:643]: changed, let me know if it fits better for you

File: pkg/controller/plan/adapter/base/doc.go [hunk 22]
```diff
@@ -51,6 +52,8 @@ type Builder interface {
 	ResolvePersistentVolumeClaimIdentifier(pvc *core.PersistentVolumeClaim) string
 	// Conversion Pod environment
 	PodEnvironment(vmRef ref.Ref, sourceSecret *core.Secret) (env []core.EnvVar, err error)
+	// Build a PersistentVolumeClaim with SourceRef for VolumePopulator
+	PersistentVolumeClaimWithSourceRef(da ovirt.XDiskAttachment, storageName *string, populatorName string, accessModes []core.PersistentVolumeAccessMode, volumeMode *core.PersistentVolumeMode) *core.PersistentVolumeClaim
```
[the remark this answers is not in the document]
Comment (liranr23) [reply pkg/controller/plan/adapter/base/doc.go:56]: because `StorageClassName *string `json:"storageClassName,omitempty" protobuf:"bytes,5,opt,name=storageClassName"`` we can do non-pointer and then use `&` as well..

[the remark this answers is not in the document]
Comment (ahadas) [reply pkg/controller/plan/adapter/base/doc.go:56]: nah, if it's that way on the source you take the value from, that's ok (I wanted to see if we can minimize the length of the arguments list a bit)

File: pkg/controller/plan/kubevirt.go [hunk 23]
```diff
@@ -555,11 +573,131 @@ func (r *KubeVirt) getPVCs(vm *plan.VMStatus) (pvcs []core.PersistentVolumeClaim
...
+						break
+					}
+				}
+			}
+		}
+	}
+	return
+}
+
+func (r *KubeVirt) createVolumes(vm ref.Ref) (err error) {
```
Comment (nyoxi) [inline pkg/controller/plan/kubevirt.go:595]: This seems oVirt specific. Could the name reflect that?

File: pkg/controller/plan/kubevirt.go [hunk 24]
```diff
@@ -555,11 +573,131 @@ func (r *KubeVirt) getPVCs(vm *plan.VMStatus) (pvcs []core.PersistentVolumeClaim
...
+		if failure != nil && !k8serr.IsAlreadyExists(failure) {
+			return failure
+		}
+
+		accessModes, volumeMode, failure := r.getStorageProfileModes(*storageName)
+		if failure != nil {
+			return failure
+		}
+
+		pvc := r.Builder.PersistentVolumeClaimWithSourceRef(da, storageName, populatorCr.Name, accessModes, volumeMode)
```
Comment (nyoxi) [inline pkg/controller/plan/kubevirt.go:619]: Check for `nil` would be useful, although it cannot happen with current state of the code.

File: pkg/controller/plan/adapter/ovirt/builder.go [hunk 25]
```diff
@@ -412,20 +414,13 @@ func (r *Builder) mapDisks(vm *model.Workload, persistentVolumeClaims []core.Per
...
-	for i := range persistentVolumeClaims {
-		pvc := &persistentVolumeClaims[i]
-		pvcMap[r.ResolvePersistentVolumeClaimIdentifier(pvc)] = pvc
-	}
-
-	for i, da := range vm.DiskAttachments {
-		pvc := pvcMap[da.Disk.ID]
-		volumeName := fmt.Sprintf("vol-%v", i)
+	for _, da := range vm.DiskAttachments {
+		volumeName := da.Disk.ID
```
[the remark this answers is not in the document]
Comment (yaacov) [reply pkg/controller/plan/adapter/ovirt/builder.go:418]: no, ( AFAIK :-) )

File: cmd/ovirt-populator/ovirt-populator.go [hunk 26]
```diff
@@ -103,6 +95,7 @@ func populate(crName, engineURL, secretName, diskID, volPath, namespace, crNames
 	}
 
 	gvr := schema.GroupVersionResource{Group: groupName, Version: apiVersion, Resource: resource}
+	*/
```
Comment (ahadas) [inline cmd/ovirt-populator/ovirt-populator.go:98]: drop then?

File: cmd/ovirt-populator/ovirt-populator.go [hunk 27]
```diff
@@ -119,7 +112,7 @@ func populate(crName, engineURL, secretName, diskID, volPath, namespace, crNames
 				klog.Error(err)
 			}
 
-			if progressOutput.Size != nil {
+			/*if progressOutput.Size != nil {
```
Comment (ahadas) [inline cmd/ovirt-populator/ovirt-populator.go:115]: please add TODO

File: cmd/ovirt-populator/ovirt-populator.go [hunk 28]
```diff
@@ -166,12 +159,24 @@ func loadEngineConfig(secretName, engineURL, namespace string) engineConfig {
 	secret, err := clientset.CoreV1().Secrets(namespace).Get(context.TODO(), secretName, metav1.GetOptions{})
 	if err != nil {
 		klog.Fatal(err.Error())
+	}*/
```
Comment (ahadas) [inline cmd/ovirt-populator/ovirt-populator.go:162]: drop?

File: cmd/ovirt-populator/ovirt-populator.go [hunk 29]
```diff
@@ -166,12 +159,24 @@ func loadEngineConfig(secretName, engineURL, namespace string) engineConfig {
...
+	user, err := os.ReadFile("/etc/secret-volume/user")
+	if err != nil {
+		klog.Error(err.Error())
```
Comment (ahadas) [inline cmd/ovirt-populator/ovirt-populator.go:165]: missing 'return' here and below?

File: cmd/ovirt-populator/ovirt-populator.go [hunk 30]
```diff
@@ -92,7 +84,7 @@ func populate(crName, engineURL, secretName, diskID, volPath, namespace, crNames
 		diskID,
 		volPath,
 	}
-	config, err := rest.InClusterConfig()
+	/*config, err := rest.InClusterConfig()
```
Comment (liranr23) [inline cmd/ovirt-populator/ovirt-populator.go:87]: many things are comment out now. do we wish to preserve it with a comment? or we should just drop it?
mostly talking about the progress.

File: pkg/controller/plan/kubevirt.go [hunk 31]
```diff
@@ -555,11 +573,138 @@ func (r *KubeVirt) getPVCs(vm *plan.VMStatus) (pvcs []core.PersistentVolumeClaim
...
+		if pvc == nil {
+			return
```
Comment (ahadas) [inline pkg/controller/plan/kubevirt.go:625]: how about adding a log?

**Comment by liranr23** [comment]
CI won't work because it uses kubev2v populator images.

File: pkg/controller/plan/kubevirt.go (hunk 18 above)
Comment (ahadas) [reply pkg/controller/plan/kubevirt.go]: it was renamed to arePVCsReady which is better but I think the name should reflect that it only handles PVCs for oVirt

**Review by ahadas — COMMENTED** [review]
much better!

File: .github/workflows/pull-request.yml [hunk 32]
```diff
@@ -77,6 +77,12 @@ jobs:
           kubectl logs -n konveyor-forklift $(kubectl get po -n konveyor-forklift  -o=name | grep forklift-controller) >> /tmp/artifacts/k8s-forklift-controller-inventory.log
           kubectl get Storageclasses -A -o yaml >> /tmp/artifacts/k8s-storage-classes.log
           
+          if [ "${{ matrix.source_provider }}" == 'ovirt' ] ; then
```
Comment (ahadas) [inline .github/workflows/pull-request.yml:80]: @eslutsky why only to ovirt?

**Review by ahadas — COMMENTED** [review]
minor comment about the logs we take in CI, other than that - lgtm

File: vendor/github.com/kubev2v/lib-volume-populator/populator-machinery/controller.go [hunk 33]
```diff
@@ -716,13 +726,27 @@ func (c *controller) syncPvc(ctx context.Context, key, pvcNamespace, pvcName str
...
 }
 
-func makePopulatePodSpec(pvcPrimeName string) corev1.PodSpec {
+func makePopulatePodSpec(pvcPrimeName, secretName string) corev1.PodSpec {
+	nonRoot := true
+	allowPrivilageEscalation := false
+	user := int64(1000)
 	return corev1.PodSpec{
-		ServiceAccountName: "forklift-controller",
+		//ServiceAccountName: "forklift-controller",
```
Comment (ahadas) [inline vendor/github.com/kubev2v/lib-volume-populator/populator-machinery/controller.go:734]: drop?

Comment (liranr23) [reply vendor/github.com/kubev2v/lib-volume-populator/populator-machinery/controller.go:734]: eh it will require another push to the lib. forgot it, sorry.

**Comment by liranr23** [comment]
locally the CI passes..

File: .github/workflows/pull-request.yml (hunk 32 above)
Comment (eslutsky) [reply .github/workflows/pull-request.yml]: fixed

**Comment by eslutsky** [comment]
@liranr23 , according to the logs looks like we are facing a [RBAC](https://stackoverflow.com/questions/69121227/mountvolume-setup-failed-for-volume-volume-name-token-m4rtn-failed-to-sync) issue in the CI,
ovirt-populator pod is using default serviceAccount on a generated namespace ,
from the events we can see that the pod failed to scheduled:
``` failed for volume "secret-volume" : failed to sync secret cache: timed out waiting for the condition``` 
so it must be missing some permissions.

File: .github/workflows/pull-request.yml [hunk 34]
```diff
@@ -81,6 +81,8 @@ jobs:
             kubectl get all -n $(kubectl get ns | grep forklift-e2e-tests) -o yaml >> /tmp/artifacts/k8s-all-forklift-tests-ns-objects.log
             kubectl logs -n kube-system $(kubectl get po -n kube-system  -o=name | grep csi-nfs-controller)  csi-provisioner >> /tmp/artifacts/k8s-csi-nfs-provisioner.log
             kubectl logs -n kube-system $(kubectl get po -n kube-system  -o=name | grep csi-nfs-controller)  nfs >> /tmp/artifacts/k8s-csi-nfs.log
+            kubectl get sa -n $(kubectl get ns | grep forklift-e2e-tests) -o yaml >> /tmp/artifacts/k8s-service-accounts.log
```
Comment (eslutsky) [inline .github/workflows/pull-request.yml:84]: this one wont work, you need to cut the namespace only like this:
```
kubectl get sa-n $(kubectl get ns -o=name |grep forklift-e2e-tests | cut -d/ -f2) -o yaml >> /tmp/artifacts/k8s-service-accounts.log
```

File: vendor/github.com/kubev2v/lib-volume-populator/populator-machinery/controller.go [hunk 35]
```diff
@@ -460,6 +455,28 @@ func (c *controller) syncPvc(ctx context.Context, key, pvcNamespace, pvcName str
...
+	if err != nil {
+		return err
+	}
+
+	// Overriding the namespace to our target namespace
+	// Taking the secret name from the args
+	var secretName string
+	for _, val := range args {
+		if strings.HasPrefix(val, "--cr-namespace=") {
+			c.populatorNamespace = strings.Split(val, "--cr-namespace=")[1]
```
Comment (ahadas) [inline vendor/github.com/kubev2v/lib-volume-populator/populator-machinery/controller.go:475]: why is this needed? don't we set the namespace to that of the CR?

File: vendor/github.com/kubev2v/lib-volume-populator/populator-machinery/controller.go [hunk 36]
```diff
@@ -460,6 +455,28 @@ func (c *controller) syncPvc(ctx context.Context, key, pvcNamespace, pvcName str
...
+
+	// Calculate the args for the populator pod
+	var args []string
+	args, err = c.populatorArgs(rawBlock, unstructured)
+	if err != nil {
+		return err
+	}
+
+	// Overriding the namespace to our target namespace
+	// Taking the secret name from the args
```
Comment (ahadas) [inline vendor/github.com/kubev2v/lib-volume-populator/populator-machinery/controller.go:471]: I don't think the comment about the secrets adds any value - the code is self-explanatory

File: cmd/ovirt-populator/ovirt-populator.go (hunk 29 above)
Comment (liranr23) [reply cmd/ovirt-populator/ovirt-populator.go:136]: yeah, maybe we should change to Fatal as it was - then it should crash the pod. I changed it to Error for debugging purpose

File: vendor/github.com/kubev2v/lib-volume-populator/populator-machinery/controller.go (hunk 35 above)
Comment (liranr23) [reply vendor/github.com/kubev2v/lib-volume-populator/populator-machinery/controller.go:475]: no, that's the problem. the `c.populatorNamespace` is set with the value given from the `populator-controller` and it's the namespace the `populator-controller` is running. with this line, we take the namespace from the `OvirtVolumePopulator`.

Comment (ahadas) [reply vendor/github.com/kubev2v/lib-volume-populator/populator-machinery/controller.go:475]: ack so can't we change in the populator-controller to provide the namespace that the CR resides in?

File: .github/workflows/pull-request.yml (hunk 32 above)
Comment (liranr23) [reply .github/workflows/pull-request.yml]: @eslutsky i saved all the changes to a single PR locally, let me know if you want me to post it or you will.

File: vendor/github.com/kubev2v/lib-volume-populator/populator-machinery/controller.go (hunk 35 above)
Comment (ahadas) [reply vendor/github.com/kubev2v/lib-volume-populator/populator-machinery/controller.go:473]: ok, I missed that it's not part of the populator pod but the populator controller and we still have that populatorNamespace that is initialized to POD_NAMESPACE - this requires a bit more cleanup

File: cmd/ovirt-populator/ovirt-populator.go [hunk 37]
```diff
@@ -152,26 +130,27 @@ func populate(crName, engineURL, secretName, diskID, volPath, namespace, crNames
 	}
 }
 
-func loadEngineConfig(secretName, engineURL, namespace string) engineConfig {
-	config, err := rest.InClusterConfig()
+func loadEngineConfig(engineURL string) engineConfig {
+	user, err := os.ReadFile("/etc/secret-volume/user")
 	if err != nil {
 		klog.Fatal(err.Error())
+		return engineConfig{}
```
Comment (ahadas) [inline cmd/ovirt-populator/ovirt-populator.go:137]: no need for this anymore

**Review by ahadas — COMMENTED** [review]
minor comment, besides that lgtm

**Review by ahadas — APPROVED** [review]
